docs/USAGE-GUIDE.md

```diff
@@ -1929,6 +1929,95 @@ s2n-tls will consider all allocated memory available for use, and will attempt t
+
+Once **s2n_recv_early_data** has been called once, it must be called until it returns success, indicating that all early data has been read. If an application stops calling it early, some early data may be left unread and cause later calls to **s2n_negotiate** to fail and close the connection with a protocol error. Calling **s2n_recv_early_data** again after it returns success is possible but does nothing.
+
+Once a server has read all early data, you should call **s2n_negotiate** to complete the handshake just as you would for a handshake that did not include early data.
+
+For example:
+```
+uint8_t early_data[MAX_EARLY_DATA] = { 0 };
+ssize_t total_data_recv = 0, data_recv = 0;
+while (s2n_recv_early_data(conn, early_data + total_data_recv, MAX_EARLY_DATA - total_data_recv,
+        &data_recv, &blocked) != S2N_SUCCESS) {
+    total_data_recv += data_recv;
+    if (s2n_error_get_type(s2n_errno) != S2N_ERR_T_BLOCKED) {
+        exit(1);
+    }
+}
+while (s2n_negotiate(conn, &blocked) != S2N_SUCCESS) {
+    if (s2n_error_get_type(s2n_errno) != S2N_ERR_T_BLOCKED) {
+        exit(1);
+    }
+}
+```
+
+### Adding anti-replay protection
+s2n-tls does not include anti-replay protection automatically. Effective anti-replay protection for a multi-server application requires an external state shared by all servers. Without shared state, an attacker can capture early data originally sent to server A and successfully replay it against server B.
+
+Instead, s2n-tls provides a callback that you can implement to hook into your anti-replay solution. The **s2n_early_data_cb** can be used to accept or reject client requests to use early data. The callback can be configured by using **s2n_config_set_early_data_cb**.
+
+Using the **s2n_offered_early_data** pointer offered by the callback, **s2n_offered_early_data_reject** and **s2n_offered_early_data_accept** can be called to accept or reject the client request to use early data. The callback can be implemented asynchronously by returning without calling either method, causing the connection to block until the early data is either rejected or accepted.
+
+Like most s2n-tls callbacks, **s2n_early_data_cb** takes an optional context argument to allow custom state to be passed to implementations of the callback. Unlike most s2n-tls callbacks, the optional context is not configured when the callback is set. Instead, the context is associated with the specific pre-shared key or session ticket used for early data. The context can be set for external pre-shared keys by calling **s2n_psk_set_early_data_context**. For session tickets, **s2n_connection_set_server_early_data_context** can be used to set the context that a server includes on its new session tickets. Because the session ticket version of the context needs to be serialized into the ticket, the early data context is a byte buffer instead of the usual void pointer
```

Comment:
Probably just remove "Like most s2n-tls callbacks,". It's superfluous. 

"to be passed to implementations of the callback" --> "to be passed to the callback."

I prefer: "...can be used to set the context the server includes on its new session tickets"

Finally, you need a period at the end of this paragraph.